R/github-pat.R

```diff
@@ -44,19 +69,29 @@ set_github_pat <- function(force_new = FALSE, validate = interactive(), verbose
       }
       if(isTRUE(validate)) {
         hx <- curl::handle_setheaders(curl::new_handle(), Authorization = paste("token", cred$password))
-        req <- curl::curl_fetch_memory("https://api.github.com/user", handle = hx)
+        host <- parse_url(url)["host"]
+        api_url <- switch(
+          host,
+          `github.com` = "https://api.github.com/user",
+          sprintf("https://%s/api/v3/user", host)
+        )
+        req <- curl::curl_fetch_memory(api_url, handle = hx)
         if(req$status_code >= 400){
+          # TODO: ideally we would reveal more about the error; we have the info
           message("Authentication failed. Token invalid.")
           credential_reject(cred)
           next
         }
         if(verbose == TRUE){
           data <- jsonlite::fromJSON(rawToChar(req$content))
           helper <- tryCatch(credential_helper_get()[1], error = function(e){"??"})
-          message(sprintf("Using GITHUB_PAT from %s (credential helper: %s)", data$name, helper))
+          message(sprintf(
+            "Using GITHUB_PAT for %s@%s (credential helper: %s)",
+            data$login, cred$host, helper
+          ))
```

Comment:
Things to note:

* I'm revealing the `host` and `username` (instead of "wallet name").
* I think `data$login` and `cred$username` are the same? Which brings up: it seems that `git-credential` does some initial remote call? Because somehow, after the initial `"fill"` command, `cred` gains a `username` component and it always seems to be current. Do you know more about that? I guess it still makes sense to make an explicit API call to check validity?

Reply:
Hmm this sounds very mysterious to me, afaik `git-credential` does no such thing. As I understand it, the caller needs to do the validation and then call `credential_approve` to say if the credential was valid or not.

I really don't understand how it can know your usename if you only gave it a PAT.

Reply:
> I really don't understand how it can know your username if you only gave it a PAT.

I agree and I read their docs saying the caller has to validate the token. But I just know that when I cleared the keychain and I put a browser statement here:

https://github.com/r-lib/credentials/blob/934bea297a2433bb679d6da21d203c6ea817f9e5/R/http-credential.R#L34-L35

and entered a token at the prompt, the returned `cred` already has a `username`.

```
> set_github_pat("https://jennybryan@github.com")
If prompted for GitHub credentials, enter your PAT in the password field
Called from: git_credential_ask(url, verbose = verbose)
Browse[1]> cred
    username     protocol         host 
"jennybryan"      "https" "github.com" 
```

And this is prior to the API call that credentials does for validation.

Reply:
Sorry let my try that again without providing a username myself. Let me switch back to `master` where I first observed this.

Reply:
You're right. Disregard my speculation that `git-credential fill` does anything re: determining username.

Reply:
Main proposal here is to include GitHub username instead of "wallet name".